OCPEDGE-1318: Filled out remaining sections and updates for fencing.

jaypoulz opened the pull request:
This is a collection of updates which attempts to clean up some of the
remaining template-boiler plate of the enhancement.

I've also added sections to explain recent discussions, like:
- Baremetal Platform vs. None
- Fencing initialization via an in-cluster operator
- Why the Assisted Installer SaaS is probably not compatible
- Whether or note in-cluster networking is compatible

jaypoulz committed:
OCPEDGE-1318: Filled out remaining sections and updates for fencing.

This is a collection of updates which attempts to clean up some of the
remaining template-boiler plate of the enhancement.

I've also added sections to explain recent discussions, like:
- Baremetal Platform vs. None
- Fencing initialization via an in-cluster operator
- Why the Assisted Installer SaaS is probably not compatible
- Whether or note in-cluster networking is compatible

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
+3. Enabling fencing in RHEL-HA
 
-Otherwise, the procedure follows the standard flow except for the configuration of 2 nodes instead of 3. At this time we've discussed the collection of RedFish details (including passwords!) for each node. This is needed for the RHEL-HA configuration by leveraging the BareMetalHost CRDs populated from the baremetal platform specification in the install-config. There are open questions on how to e ...
+We propose the inclusion of a new in-cluster operator responsible for the remediation of 2NO resource agents.
```
I'm not sure I understand this part.
- Even though I'm pro on a new operator IIUC the decision was we are trying to embed the functionality in existing operators .
- I'm struggling with the term `remediation` here, I'm not sure it's appropriate in this context. The way I understand it, the operator only sets up the resource agents and the agents are responsible for the fencing/remediation

jaypoulz replied:
As discussed in our 1/1 this morning, my prediction was that the outcome of today's 2NO sync would be that we'd reestablish the desire for an in-cluster operator to handle the fencing setup.

If the in-cluster operator was responsible for initializing  pacemaker and making sure its resource agents were running healthily, my thought was that this makes the operator's job management of pacemaker resource agents. (i.e. pacemaker remediation, not cluster remediation)

My prediction was wrong. :) we are instead looking at the simple flow of just using MCO to stick a file on disk, so my new draft will remove all references to this operator (outside of the parts that talk about it being an option in case we need it for initializing pacemaker).

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
-Fencing setup is the last important aspect of the cluster installation. In order for the cluster installation to be successful, fencing should be configured and active before we declare the installation successful. Ideally, the fencing secrets should be made available to the control-plane nodes in the initial pacemaker initialization so that fencing can be configured during step 2. There are a fe ...
-1. Should fencing be made active during the installation or should pacemaker start with it disabled and only enable it after being signaled by the in-cluster entity when the cluster installation is detected as successful?
-2. What mechanism will pacemaker use to get access to the secret linked from the BareMetalHost CRD?
+###### Configuring Fencing Via 2NO Resource Agent Remediation Operator
```
Nit: I might consider a more general name for the operator, maybe 2no operator or 2no setup operator.

jaypoulz replied:
Yes - we're on the same page based on today's discussion. Any references to the operator will be simplified to indicate it's just for setup.

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
-Everything else about cluster creation will be an opaque implementation detail not exposed to the user.
+For a two-node cluster to be successful, we need to ensure the following:
+1. The new operator must be able to watch for changes to the new CRD and synchronize the BMC secrets for RHEL-HA once during initialization and again whenever the CRD is updated.
+3. The new operator should run the `pcs` configuration command on the nodes to enable the fencing agent and report an error to the user if the fencing configuration isn't successful. In the case of installation, this means that the operator becomes degraded to ensure that the installation fails.
```
looks like number ordering was  messed up during the edit

jaypoulz replied:
Good catch :D

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
 
-Everything else about cluster creation will be an opaque implementation detail not exposed to the user.
+For a two-node cluster to be successful, we need to ensure the following:
+1. The new operator must be able to watch for changes to the new CRD and synchronize the BMC secrets for RHEL-HA once during initialization and again whenever the CRD is updated.
```
I might be wrong, but IIRC this part will is already covered by pacemaker (i.e making sure BMC credentials are synced) 
@fabbione please keep me honest here :sweat_smile:

fabbione replied:
@mshitrit it depends how fencing is configured in pacemaker itself. Usually there 3 ways to manage fencing credentials in pacemaker for fencing.

1. username/password are given to pacemaker via pcs configuration. In this case if any of those are changed, a pcs config update is required. pacemaker will take care to synchronize the configuration change across all nodes of the cluster.
2. password is stored in a file on disk. In this case the operator will update the file and pacemaker/fence agent will pick it up automatically on the next run.
3. password-script, a shell/python/$interpreter script that will get the credentials somehow/somewhere. Same as point 2, it will be executed each time a fencing operation is required.

For now, we should stick to either 1 or 2. Grabbing credentials in 3 can be challenging depending on failure scenarios.

jaypoulz replied:
Thanks for breaking this down @fabbione.
For this draft, I'm going to assume we're following option 2. We use MCO to stick the credentials on disk, and if an operation needs to update that password file, it can be updated by an MCO config or manually logging in. I will make a note in the open questions describing the alternative for option 1 in case someone raises an issue with that flow.

jaypoulz replied:
I decided to go into less detail than I originally intended. Just noted that the secrets are on disk via MCO to be consumed by pacemaker however we decide to do that.

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
+For a two-node cluster to be successful, we need to ensure the following:
+1. The new operator must be able to watch for changes to the new CRD and synchronize the BMC secrets for RHEL-HA once during initialization and again whenever the CRD is updated.
+3. The new operator should run the `pcs` configuration command on the nodes to enable the fencing agent and report an error to the user if the fencing configuration isn't successful. In the case of installation, this means that the operator becomes degraded to ensure that the installation fails.
+4. The new operator should periodically check that the fencing agent is healthy to ensure that the user can be notified in case the credential is rotated and needs to be updated.
```
I think this has some major implications I'm not sure we've discussed.
The two things from the top of my mind are:
- The 2no operator will need to be running beyond the setup scope
- We assume fence agents needs external monitoring

fabbione replied:
@mshitrit pacemaker does monitoring of the fencing devices on a regular base to report to the administrator that something is off.

However, it is incorrect to assume that the monitor operation will fail only because credentials needs to be updated.

A monitoring operation can fail for multiple reasons, including an admin performing a BMC firmware update that don´t need a server reboot, or a network blip between the host and the BMC, etc. etc.

jaypoulz replied:
Based on today's discussion, monitoring will be primarily driven through pacemaker itself. It remains an open discussion on the best way to raise this to a higher level component (i.e. how to get the cluster to tell the user that something is wrong). This can probably be done via monitoring, but this needs its own investigation.

mshitrit commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -300,26 +369,29 @@ This proposal is an alternative architecture to Single-node and MicroShift, so i
+   4. Mitigation: It may be possible to identify scenarios where, for a known hardware topology, it is safe to allow the node to proceed automatically.
 
-1. Risk: “Something changed, let's reboot” is somewhat baked into OCP’s DNA and has the potential to be problematic when nodes are actively watching for their peer to disappear, and have an obligation to promptly act on that disappearance by power cycling them.
+5. Risk: “Something changed, let's reboot” is somewhat baked into OCP’s DNA and has the potential to be problematic when nodes are actively watching for their peer to disappear, and have an obligation to promptly act on that disappearance by power cycling them.
```
Might worth mentioning that most of those events (maybe even all of them :thinking: ?) falls into the [first bucket](https://github.com/mshitrit/enhancements/pull/3/files#diff-f4f2dd40af3dfd355440cc5094e2c4d21877db4e5cd0d1c6830ad7b9eb84186cR275) mentioned, which means it's already handled.

eranco74 commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
+
+Otherwise, the procedure follows the standard flow except for the configuration of 2 nodes instead of 3.
+
+Because BMC passwords are being collected to initialize fencing, the SaaS offering will not be available (to avoid storing customer BMC credentials in a Red Hat database). ZTP may be considered in the future, but this will need further evaluation to ensure passwords are appropriately handled.
```
Just a note regarding ZTP, it works today with the BMC credentials, no reason it would be any different for 2NO

jaypoulz replied:
I think I'm just going to simplify this to `ZTP may be considered in the future.` because I haven't seen any technical reason for why we should or shouldn't do it (and I don't know enough about it to speak to it deeper than that).

eranco74 commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
 
-User creation of a two-node control-plane will be possible via the Assisted Installer. A key requirement is that the cluster can be deployed using only 2 nodes because requiring a third baremetal server for installation can be expensive when deploying baremetal at scale. To accomplish this, deployments will take advantage of the Assisted Installer's ability to use one of the target machines as th ...
+User creation of a two-node control-plane is possible via the Assisted Installer and the Agent-Based Installer (ABI). The initial implementation will focus on providing support for the Assisted Installer in managed cluster environments (i.e. ACM), followed by stand-alone cluster support via the Agent-Based Installer.
+The requirement that the cluster can be deployed using only 2 nodes is key because requiring a third baremetal server for installation can be expensive when deploying baremetal at scale. To accomplish this, deployments will use one of the target machines as the bootstrap node before it is rebooted into a control-plane node.
```
Would it be possible to install 2NO using the openshift-installer? (even if it requires a 3rd auxiliary node during the cluster installation)

jaypoulz replied:
Yes, but our intent is not to support this. Maybe I'll make a note about this.

eranco74 commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
-An important facility of the installation flow is the transition from a CEO deployed etcd to one controlled by pacemaker. The basic transition works as follows:
+###### Transitioning Etcd Management to RHEL-HA
+An important facility of the installation flow is the transition from a CEO deployed etcd to one controlled by RHEL-HA. The basic transition works as follows:
 1. MCO Extensions are used to ensure that pacemaker, corosync, and resource agents are pre-configured on CoreOS using installation manifests.
```
Can you elaborate on `MCO Extensions` and `installation manifests`?
Does it mean that MCO will generate these manifests (similar to [this](https://github.com/openshift/machine-config-operator/blob/67789532e4569ece95eb98fed1dde03409f065e9/pkg/operator/bootstrap.go#L185)) or we expect the installer / user to provide them as [extraManifests](https://github.com/openshift/installer/blob/2ea001d6d7cb97b1eb081c80b48b1151f2f30257/pkg/asset/agent/manifests/extramanifests.go#L16) for MCO to apply them

jaypoulz replied:
I will provide some supportive links, but what I mean by mco extensions is this:
https://docs.openshift.com/container-platform/4.17/machine_configuration/machine-configs-configure.html#rhcos-add-extensions_machine-configs-configure
https://github.com/openshift/installer/blob/2ea001d6d7cb97b1eb081c80b48b1151f2f30257/docs/user/customization.md?plain=1#L561

In short - we need to install additional RPMs on top of RHCOS. MCO extensions are a feature that allows the user to install a curated set of packages on top of RHCOS. For the pacemaker & corosync rpms, I think the answer is very much that MCO does that. However, we also need to install some configuration logic for our 2node specific pacemaker. That isn't packaged as an RPM, so I guess it could be either.

@clobrano - any input?

clobrano replied:
> we need to install additional RPMs on top of RHCOS...

that's correct.

> I think the answer is very much that MCO does that

I don't know the details, but if MCO can understand its 2NO by the `platform`, then I tend to agree, otherwise would be the installer responsibility (not the user though)

> However, we also need to install some configuration logic for our 2node specific pacemaker. That isn't packaged as an RPM, so I guess it could be either.

This are MachineConfigs we can install via the operator. Whether it's a custom operator for 2NO (like the one we have for the current demo) or the extension of an already existing operator.

jaypoulz replied:
I think I captured this. We may want to review the wording in the newest patch.

eranco74 commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -113,32 +119,57 @@ In both cases, the control-plane's dependence on etcd will cause it to respond w
-2. Upon detection that the cluster infrastructure is using the DualReplica controlPlaneTopology in the infrastructure config, an in-cluster entity (see open questions regarding whether this should be handled by CEO or an additional operator) will run a command on one of the cluster nodes to initialize pacemaker. The outcome of this is that the resource agent will be started on both nodes.
-3. The aforementioned in-cluster entity will signal CEO to relinquish control of etcd by setting CEO's `managedEtcdKind` to `External`. When this happens, CEO removes the etcd pod from the static pod configs. The resource agents for etcd are running from step 2, and they are configured to wait for etcd pods to be gone so they can restart them using Podman.
+2. Upon detection that the cluster infrastructure is using the DualReplica controlPlaneTopology in the infrastructure config, an in-cluster entity (see open questions regarding whether this should be handled by CEO or the proposed 2NO resource agent operator) will run a command on one of the cluster nodes to initialize pacemaker. The outcome of this is that the resource agent will be started on b ...
+3. The aforementioned in-cluster entity will signal CEO to relinquish control of etcd by setting CEO's `managedEtcdKind` to `External`. When this happens, CEO immediately removes the etcd pod from the static pod configs. The resource agents for etcd are running from step 2, and they are configured to wait for etcd pods to be gone so they can restart them using Podman.
```
What happens if the etcd pod definition need to change after etcd is started using Podman, who is in charge for updating and restarting it?

tjungblu replied:
the same holds also for certificates - I would leave this for later (GA).

jaypoulz replied:
Yes. :)

That is a good question. I will add a note for this.

clobrano replied:
I see two cases
* image change (kind of frequent)
* configuration change (hopefully not frequent, but possible)

in case of image update, there's already a planned change to get the podman image from the pod definition, instead than from the input at agent creation time, like we do now.

for configuration changes, we need something capable of mapping pod definition to container. A good candidate is [podman play kube](https://docs.podman.io/en/v4.2/markdown/podman-play-kube.1.html)

jaypoulz replied:
Do we still need etcd to render the static pod definition, but not the the static pod directory?

jaypoulz replied:
I've updated this section to with some of the background from this thread.

clobrano replied:
> Do we still need etcd to render the static pod definition, but not the the static pod directory?

I don't understand the question, can you clarify?

jaypoulz replied:
My idea is that since CEO would normally roll out updates to etcd, it would normally push those updates out as changes to the static pod definition. Instead of doing that, what we might be able to do is have etcd still render that the block related to the etcd pod, except that it would write it to disk somewhere other than the static pod location. Then, all pacemaker has to do is monitor changes to the CEO-updated file and redeploy whenever that file is touched.

clobrano replied:
if the update refers to image, we can make the agent able to pick it automatically (e.g. looking into the image definition of `etcdctl` container, which uses the same image as `etcd`) at each startup, so the image would be always up to date.
For different changes, we can define categories that the agent will be able to "understand" (for example I'm thinking about the value of some environment variables it uses)

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -91,17 +97,17 @@ This requires our solution to provide a management experience consistent with "n
-Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to deliver supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
-etcd will run as as a voting member on both nodes.
+We will use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to deliver supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+Etcd will run as a a voting member on both nodes.
```
etcd is correct https://etcd.io/

jaypoulz replied:
I will rephrase this sentence so it doesn't conflict with english grammar rules :)

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -178,29 +210,37 @@ This will be achieved by having the same entity that drives the configuration of
 To enable this flow, we propose the addition of a `managedEtcdKind` field which defaults to `Cluster` but will be set to `External` during installation, and will only be respected if the `Infrastructure` CR's `TopologyMode` is `DualReplicaTopologyMode`.
 This will allow the use of a credential scoped to `ConfigMap`s in the `openshift-etcd-operator` namespace, to make the change.
```
don't you guys want to remove this "configmap" story above?

jaypoulz replied:
I think the pacemaker initialization script is still what sets this `managedEtcdKind` to external. If we kick off that script from CEO, we could rework it so that CEO does this step, but IIUC the current intent is for pacemaker to kick off this change.

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -178,29 +210,37 @@ This will be achieved by having the same entity that drives the configuration of
 A switch to [MCO Layering](../ocp-coreos-layering/ocp-coreos-layering.md ) will be investigated once it is GA in a shipping version of OpenShift.
 
-Once installed, the configuration of the RHEL-HA components will be done via an in-cluster entity. This entity could be a dedicated in-cluster operator or a function of CEO triggering a script on one of the control-plane nodes. This initialization will require that RedFish details have been collected by the installer.
+Once installed, the configuration of the RHEL-HA components will be done via an in-cluster entity. This entity could be a dedicated in-cluster operator or a function of CEO triggering a script on one of the control-plane nodes.
```
I think CEO will not be able to run any script on a host:
https://github.com/openshift/cluster-etcd-operator/blob/master/manifests/0000_20_etcd-operator_06_deployment.yaml#L28-L29

also not sure that the security team is happy to give another operator root privs

jaypoulz replied:
This may be one of the factors that necessitates a new 2no-setup operator. Either way, an operator needs privileges to run the setup script. If it's not CEO, that it's still something. i'll try to capture that here

jaypoulz replied:
Added a note to cover the risk of privilege escalation.

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -212,9 +252,38 @@ More information on creating OCF agents can be found in the upstream [developer
 
-As part of the fencing setup, the cri-o and kubelet services will still be owned by systemd when running under pacemaker. The main difference is that the resource agent will be responsible for signaling systemd to change their active states. The etcd pods are different in this respect since they will be restarted using Podman. It may be possible to start these with the same user account as the or ...
+As part of the fencing setup, the cri-o and kubelet services will still be owned by systemd when running under pacemaker. The main difference is that the resource agent will be responsible for signaling systemd to change their active states.
+The etcd pods are different in this respect since they will be restarted using Podman, but this will be running as root, as it was under CEO.
```
only the etcd _container_ is running under podman. I wonder, would it still require root? I would assume so, because the /var/lib/etcd folder would otherwise need a chmod

jaypoulz replied:
The note about running as root was to help address a concern that the process updates changing which user runs the process could have implications like thread limitations.

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -300,26 +369,29 @@ This proposal is an alternative architecture to Single-node and MicroShift, so i
+   2. Mitigation: Peers remain inert (unable to fence peers, or start cri-o, kubelet, or etcd) after rebooting until they can contact their peer
 
-1. Risk: Multiple entities (RHEL-HA, CEO) attempting to manage etcd membership would cause an internal split-brain
+2. Risk: Multiple entities (RHEL-HA, CEO) attempting to manage etcd membership would cause an internal split-brain
```
it would not create a split-brain, it would just simply create multiple (similar) containers that would try to listen on the same port and run on the same database files

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -439,38 +494,51 @@ See template for guidelines/instructions.
 - Describe the possible failure modes of the API extensions.
 
+  There shouldn't be any failures introduced by adding a new topology.
+  Etcd transitioning from CEO-managed to externally managed should be a one-way switch. If it fails, it should result in a failed installation.
```
you can/should also encode this behavior into the API, as in setting to `etcd.managementState = External` would be a latch and can't be changed from there

jaypoulz replied:
I think Joel mentioned we'd do this via a ValidatingAdmissionPolicy. May I can mention that here.

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -439,38 +494,51 @@ See template for guidelines/instructions.
+
+  As mentioned above, a BMC credential rotation could result in a customer "breaking" pacemaker's ability to fence nodes. On its own, pacemaker has no way of communicating this kind
+  of failure to the cluster admin. The proposed 2NO resource agent remediation operator could detect this and warn the cluster administrator that action is required. Alternatively,
+  RHEL-HA could detect this and immediately respond by turning off etcd for one of the nodes - forcing the cluster into read-only mode until a user manually logs in and restores the fencing agent.
```
btw, OpenShift has no concept of "read-only". Since etcd and apiserver requires quorum for both writes and (linearized) reads, it will never return anything and your cluster will be down. 

Workload still continues to run.

jaypoulz replied:
This part needs to be updated based on today's discussion since pacemaker don't have an active role in this. it just let's an already broken cluster stay broken because it can proceed with fencing.

tjungblu commented on enhancements/two-nodes-openshift/2no.md:
```diff
@@ -439,38 +494,51 @@ See template for guidelines/instructions.
   and add them as reviewers to this enhancement.
 
-## Support Procedures
+  In case of escalation, the most likely team affected outside of Edge Enablement (or whoever owns the proposed topology) is the etcd team because forcing the cluster into a read-only
```
there is no etcd team anymore, it's just the "control plane team" now

tjungblu replied:
and same as the above holds, there is no read-only mode :)

jaypoulz replied:
Removed all references to read-only :)

jaypoulz committed:
Enhancement updates around remaining open questions and selected designs.

- Added explanations around which installers may be evaluated for further business cases
- Added examples for platform extensions for providing fencing credentials.
- Removed all references to pacemaker remediation operator.
- Added a section explaining why we moved away from a operator-based design.
- Noted open question around how to inform user of cluster health risk related to pacemaker.
- Noted open question on how to reconcile updates to etcd pod
- Made sure etcd is always lower case
- Update `none` vs `baremetal` to explain why we should target both
